File: algorithm_2019/range-sum-of-bst.py
"""
给定二叉搜索树的根结点 root，返回 L 和 R（含）之间的所有结点的值的和。

二叉搜索树保证具有唯一的值。


示例 1：

输入：root = [10,5,15,3,7,null,18], L = 7, R = 15
输出：32
示例 2：

输入：root = [10,5,15,3,7,13,18,1,null,6], L = 6, R = 10
输出：23
"""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
        if not root:
            return 0
        if root.val < L:
            res = self.rangeSumBST(root.right, L, R)
        elif root.val > R:
            res = self.rangeSumBST(root.left, L, R)
        else:
            left_res = self.rangeSumBST(root.left, L, root.val)
            right_res = self.rangeSumBST(root.right, root.val, R)
            res = left_res + right_res + root.val
        return res

File: Ruby/range-sum-of-bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} l
# @param {Integer} r
# @return {Integer}
def traverse_sum(node, l, r)
  if node.nil?
    return 0
  end
  if node.val < l
    sub_left_sum = 0
    sub_right_sum = traverse_sum(node.right, l, r)
    node_val = 0
  elsif node.val > r
    sub_left_sum = traverse_sum(node.left, l, r)
    sub_right_sum = 0
    node_val = 0
  else
    sub_left_sum = traverse_sum(node.left, l, r)
    sub_right_sum = traverse_sum(node.right, l, r)
    node_val = node.val
  end
  return sub_left_sum + sub_right_sum + node_val
end

def range_sum_bst(root, l, r)
  if root.nil?
    return 0
  end
  return traverse_sum(root, l, r)
end
